Adds support for Mapzen terrain

ibesora opened the pull request:
## Launch Checklist
 - [x] briefly describe the changes in this PR
Implements #5824 and adds support for the Mapzen terrarium terrain format by adding a new encoding parameter to the raster-dem source type that selects which unpack function should be used.
Also adds a new debug map that shows this new terrain format.
 - [ ] write tests for all new functionality
 - [x] document any changes to public APIs
Adds a new parameter called _encoding_ to the raster-dem source type. Defaults to mapbox to use the mapbox terrain encoding
 - [ ] post benchmark scores
 - [x] manually test the debug page
Works as expected

ibesora commented on src/style-spec/reference/v8.json:
```diff
@@ -231,7 +231,7 @@
       "values": {
         "raster-dem": {
-            "doc": "A raster DEM source using Mapbox Terrain RGB"
+            "doc": "A raster DEM source"
```
Corrected it in the new commit. 
I also expanded the declaration of the encoding parameter to explicitly tell the supported values

mollymerp commented on src/data/dem_data.js:
```diff
@@ -57,22 +57,14 @@ class DEMData {
-                level.set(x, y, this.scale * ((pixels[j] * 256 * 256 + pixels[j + 1] * 256.0 + pixels[j + 2]) / 10.0 - 10000.0));
-            }
-        }
+        this.unpackData(level, data, encoding);
```
needs to be `data.data`

mollymerp commented on src/data/dem_data.js:
```diff
@@ -57,22 +57,14 @@ class DEMData {
 
         // Build level 0
         const level = this.level = new Level(data.width, data.width / 2);
         const pixels = data.data;
```
can remove `pixels` or use it in L67

mollymerp commented on src/style-spec/reference/v8.json:
```diff
@@ -272,6 +272,10 @@
       "doc": "Contains an attribution to be displayed when the map is shown to a user."
     },
+    "encoding": {
+      "type": "string",
```
this should be `"type": "enum"` with a `values` object listing the possible types.

mollymerp commented on src/source/raster_dem_tile_source.js:
```diff
@@ -15,11 +15,14 @@ import type {Callback} from '../types/callback';
         this.type = 'raster-dem';
         this.maxzoom = 22;
         this._options = util.extend({}, options);
+        this.encoding = "mapbox";
```
`options.encoding || "mapbox";`

mollymerp commented on src/data/dem_data.js:
```diff
@@ -95,6 +87,40 @@ class DEMData {
+            for (let x = 0; x < level.dim; x++) {
+                const i = y * level.dim + x;
+                const j = i * 4;
+                // decoding per https://mapzen.com/documentation/terrain-tiles/formats/#types-of-terrain-tiles
```
link to AWS docs

ibesora commented:
Hi @mollymerp . 
When looking at your fixes I saw that I had not pushed my last commit that corrected some of the things you told me :'(
I've pushed my changes, added the AWS endpoint, changed the encoding type to an enum and added a toggle to the debug page

mollymerp commented on src/source/raster_dem_tile_source.js:
```diff
@@ -15,11 +15,14 @@ import type {Callback} from '../types/callback';
 
 
 class RasterDEMTileSource extends RasterTileSource implements Source {
+    encoding: string;
```
you can explicitly type this as `encoding: "mapbox" | "terrarium";`

mollymerp commented on src/data/dem_data.js:
```diff
@@ -95,6 +87,40 @@ class DEMData {
         this.loaded = true;
     }
 
+    unpackData(level: Level, pixels: Uint8Array | Uint8ClampedArray, encoding: string) {
```
to reduce code duplication, you could refactor to something like
```js
unpackMapbox(x, y, r, g, b) {
  level.set(x, y, this.scale * ((r * 256 * 256 +g * 256.0 + b) / 10.0 - 10000.0));
}
unpackTerrarium(x, y, r, g, b) {
  level.set(x, y, this.scale * ((r * 256 + g+ b / 256) - 32768.0));
}
const unpackFunctions = {"mapbox": unpackMapbox, "terrarium": unpackTerrarium};
const unpack = unpackFunctions[encoding];
 for (let y = 0; y < level.dim; y++) {
   for (let x = 0; x < level.dim; x++) {
      const i = y * level.dim + 
      const j = i * 4;
      unpack(x, y, pixels[j], pixels[j+1], pixels[j+2]);
  }
}
```

mollymerp commented on src/source/worker_source.js:
```diff
@@ -28,7 +28,8 @@ export type WorkerTileParameters = TileParameters & {
     coord: { z: number, x: number, y: number, w: number },
-    rawImageData: RGBAImage
+    rawImageData: RGBAImage,
+    encoding: string
```
explicitly type for enum values (as above)

mollymerp commented on src/style-spec/reference/v8.json:
```diff
@@ -272,6 +272,19 @@
+          "doc": "Terrarium format PNG tiles. See https://aws.amazon.com/es/public-datasets/terrain/ for more info."
+        },
+        "mapbox": {
+          "doc": "Mapbox Terrain RGB tiles. See https://www.mapbox.com/blog/terrain-rgb/ for more info."
```
I just found out we have official documentation for this now, so we can link to https://www.mapbox.com/help/access-elevation-data/#mapbox-terrain-rgb

ibesora commented:
Hi again @mollymerp I've just pushed some changes that incorporate your comments.

mollymerp commented on src/data/dem_data.js:
```diff
@@ -57,22 +57,14 @@ class DEMData {
     }
 
-    loadFromImage(data: RGBAImage) {
+    loadFromImage(data: RGBAImage, encoding: ?string) {
```
this typing can also be `"mapbox | "terrarium"` – sorry I missed it earlier!

ibesora replied:
Done

mollymerp commented on src/data/dem_data.js:
```diff
@@ -95,6 +87,26 @@ class DEMData {
         this.loaded = true;
     }
 
+    unpackMapbox(r: number, g: number, b: number) {
```
I think we can use `_unpack*` for these functions to signify that they're private to the `DEMData` class

ibesora replied:
Done!

mollymerp commented:
@ibesora thanks for your patience and great work here! to save you the trouble I have rebased and fixed conflicts and added some tests. I had to push to a new branch so I've made a PR for that, but I think this is good to go overall! 👉  #6110

mollymerp commented:
closing here in favor of #6110
